### ISSUE TITLE
Investigate RequirementImpl#getNamespace for EE9 (Pages/JSTL)

### ISSUE THREAD

**Issue by volosied** [OpenLiberty/open-liberty#15290]
https://github.com/OpenLiberty/open-liberty/blob/464735a8588770d8ce055a4505f53c9a029b12ab/dev/com.ibm.ws.jsp/src/com/ibm/ws/jsp/webcontainerext/JSPExtensionFactory.java#L507

The code references **com.ibm.ws.jsp.jstl.facade**, but the EE9 bundle name is **com.ibm.ws.jsp.jstl.facade._jakarta_**.

I'm not sure if any changes are needed here or how this Requirement class comes into play here. 

https://docs.osgi.org/javadoc/osgi.core/7.0.0/org/osgi/resource/Requirement.html#getNamespace--

More following up will need to be done here.

**Comment by volosied** [comment]
Spoke to Tom Watson about the return mentioned above when creating a new facade bundle (see below). 

```
**Tom Watson  1 month ago**
That should not matter as long as the new bundle continues to have this capability specified in its manifest:
Provide-Capability: com.ibm.ws.jsp.jstl.facade
```

Which explains why we haven't had any issues with it in Pages-3.0 (where. it was simply transformed).  The JSTL Refactoring PR (#15141) 
created a new facade (**[io.openliberty.jstl.facade](https://github.com/OpenLiberty/open-liberty/tree/d69348c192da4da1eade673366a6e48397eac84c/dev/io.openliberty.jstl.facade)**) for Pages-3.0 and it continues to use provide the same capability. 

As for the facade bundle, it exists to expose JSTL & JSP classes to the classloader (check for _appLoader instanceof BundleReference_) .   **Please see comment on September 23rd for more information!** 

https://github.com/OpenLiberty/open-liberty/blob/464735a8588770d8ce055a4505f53c9a029b12ab/dev/com.ibm.ws.jsp/src/com/ibm/ws/jsp/webcontainerext/JSPExtensionFactory.java#L421-L425

It also relates to APAR PI57314 -- https://www.ibm.com/support/pages/apar/PI57314

```
- adds a backwards compatibility flag for the
jspExtensionFactory (i.e., global) "osgiAppsCanProvideJSTL"
set to false by default that (when false) exposes the
standard jar packages to the jsp classloader as the parent
classloader.
- When the "osgiAppsCanProvideJSTL" flag is true, an osgi
app that uses jsps will have to either explicitly import
standard tag lib packages or include its' own copy.
```

Other than setting osgiAppsCanProvideJSTL to true,  only time this namespace is loaded is for liberty bundles (also explains why we get defects like 274796). 

I believe `osgiAppsCanProvideJSTL` is a very rare parameter as the overwhelming majority of applications use the liberty provided JSTL libraries.  Additionally, I honestly don't even how to to get Liberty to use a OSGI application -- I'm only familiar with WAR, EAR, etc. 


If my information is correct,  EE9 does not allow OSGI applications anymore. If that's the case, then we can delete this new Jakarta facade bundle and update the check above to:

`if (!osgiAppsCanProvideJSTL && (appLoader instanceof BundleReference) && LoadedPagesVersion <= 2.3) { `

Otherwise, I don't believe there is any harm is leaving everything as is it currently.

**Comment by volosied** [comment]
I attempted the changes in the PR linked above, and discovered that some bundles, which use JSPs, fail to compile. 

Some such bundle is [com.ibm.ws.security.openidconnect.server ](https://github.com/OpenLiberty/open-liberty/tree/2fd4a880754c37a988c5ed9ac4f1ea5988e465d6/dev/com.ibm.ws.security.openidconnect.server/resources)

```
[ERROR   ] Error Page Exception: 
                                                                                                               com.ibm.ws.security.oauth.2.0
                                                                                                               /oauth2
                                                                                                               Error Page Exception
                                                                                                               com.ibm.ws.jsp.webcontainerext.JSPErrorReport: JSPG0049E: /login.jsp failed to compile : 

/Users/siedlecki/libertyGit/open-liberty/dev/build.image/wlp/usr/servers/defaultServer/workarea/org.eclipse.osgi/96/data/temp/default_node/SMF_WebContainer/com.ibm.ws.security.oauth.2.0/com.ibm.ws.security.oauth.2.0/_login.java : 1 : The type javax.servlet.jsp.HttpJspPage cannot be resolved. It is indirectly referenced from required .class files

/Users/siedlecki/libertyGit/open-liberty/dev/build.image/wlp/usr/servers/defaultServer/workarea/org.eclipse.osgi/96/data/temp/default_node/SMF_WebContainer/com.ibm.ws.security.oauth.2.0/com.ibm.ws.security.oauth.2.0/_login.java : 1 : The type javax.servlet.jsp.SkipPageException cannot be resolved. It is indirectly referenced from required .class files
```

These classes were added to the facade bundle in defect 206360.  Based on how the JSP classloaders are setup (where the webapp classloader & facade bundle classloader are unified), I think we need to keep this code path open for EE9. 

_In short, I think we should leave things alone for now. Otherwise, we can place a warning or simply disable  `osgiAppsCanProvideJSTL` for EE9 and higher?_


One follow up I can do  -- Class Loaders within JSP.  I don't exactly follow the BundleWiring classloader? 

Web App Classloader -- com.ibm.ws.classloading.internal.ThreadContextClassLoaderForBundles
Facade Bundle Classloader -- org.eclipse.osgi.internal.loader.EquinoxClassLoader

I'm assuming the facade bundle exposes the JSP classes to the `com.ibm.ws.security.openidconnect.server` bundle since it doesn't import any JSP classes.

**Comment by volosied** [comment]
After an investigation into the `osgiAppsCanProvideJSTL` parameter, the facade bundles, and the inner class `RequirementImpl`, it was determined that no changes are necessary for EE9.


Roles of each:

**osgiAppsCanProvideJSTL**
- Internal property added in [PI57314](https://www.ibm.com/support/pages/apar/PI57314) 
- False by default
- When true, “an osgi app that uses jsps will have to either explicitly import standard tag lib packages or include its' own copy.”
- https://github.com/OpenLiberty/open-liberty/blob/2fd4a880754c37a988c5ed9ac4f1ea5988e465d6/dev/com.ibm.ws.jsp/resources/OSGI-INF/metatype/metatype.xml#L49-L50
- JSP Implementation Detail: whether it needs to import a standard tag library or not is due to the facade bundle.   

**Facade Bundles**
- EE8 & earlier: ----> `com.ibm.ws.jsp.jstl.facade`
- EE9 --->  `io.openliberty.jstl.facade`
- EE10 and higher will have their own bundles due to newer Tag versions. 
- The facade bundle basically imports jstl classes & jsp classes to an osgi bundle or application. 
- https://github.com/manifestlifeinc/open-liberty/blob/934b25508be3454354e54fff30165642e6a7cfbc/dev/com.ibm.ws.jsp.jstl.facade/bnd.bnd#L21-L56 
- See the attached table describing the scenarios when the facade bundle is used.
-  The way JSP uses this facade is by unifying the class loader of the facade bundle & the appLoader 
- https://github.com/OpenLiberty/open-liberty/blob/464735a8588770d8ce055a4505f53c9a029b12ab/dev/com.ibm.ws.jsp/src/com/ibm/ws/jsp/webcontainerext/JSPExtensionFactory.java#L432
- Hope the bundle description makes more sense given this explanation:
- https://github.com/OpenLiberty/open-liberty/blob/2fd4a880754c37a988c5ed9ac4f1ea5988e465d6/dev/io.openliberty.jstl.facade/bnd.bnd#L16

**RequirementImpl Class**
- This class loader unification is accomplished through a [lookup with help of various osgi classes ](https://github.com/OpenLiberty/open-liberty/blob/464735a8588770d8ce055a4505f53c9a029b12ab/dev/com.ibm.ws.jsp/src/com/ibm/ws/jsp/webcontainerext/JSPExtensionFactory.java#L423-L432)(i.e. BundleCapability, BundleWiring, etc). The facade bundle is identified by the capability it provides:
-  https://github.com/manifestlifeinc/open-liberty/blob/934b25508be3454354e54fff30165642e6a7cfbc/dev/com.ibm.ws.jsp.jstl.facade/bnd.bnd#L19 
- https://github.com/OpenLiberty/open-liberty/blob/464735a8588770d8ce055a4505f53c9a029b12ab/dev/com.ibm.ws.jsp/src/com/ibm/ws/jsp/webcontainerext/JSPExtensionFactory.java#L506-L507 

I'll be honest, I don't understand all the parts (especially classing & osgi material), but, at a high level, this explanation is adequate.
___________________
**Why nothing was changed here for EE9**

- Although Liberty EE9 does not allow osgi applications, there are still Liberty bundles which act as an osgi app. One such example is [com.ibm.ws.security.openidconnect.server ](https://github.com/OpenLiberty/open-liberty/tree/2fd4a880754c37a988c5ed9ac4f1ea5988e465d6/dev/com.ibm.ws.security.openidconnect.server/resources). The EE9 facade bundle is need to properly load the Pages/Tag libraries for this bundle . Otherwise, we get errors as those in the comment above. 
-  We could disable `osgiAppsCanProvideJSTL` for EE9, and only have be false? However, leaving it alone shouldn't have any impact due to the reason in the previous bullet (Liberty EE9 does not allow osgi applications). 
- We aren't making any breaking changes. TCK & FATs  -- all pass with the current implementation.  

**RTC Defects**
201129: Fix jsp classloading
206360: OSGI web apps with jsps require addional import-package statements even with backwards compatibility flag on


Table
<img width="715" alt="Screen Shot 2021-09-23 at 9 39 36 PM" src="https://user-images.githubusercontent.com/5934310/134607757-be613431-7fb6-4703-83b3-ae59b88ee287.png">
